File: 739. Daily Temperatures.cpp
class Solution {
public:
    struct Pair {
        int t;
        int ind;
        Pair(int temp, int index) : t(temp), ind(index) {}
    };

    vector<int> dailyTemperatures(vector<int>& temp) {
        stack<Pair> st;
        st.push(Pair(temp[0], 0));

        vector<int> ans(temp.size(), 0);

        for (int i = 1; i < (int)temp.size(); i++) {
            while (!st.empty() && temp[i] > st.top().t) {
                ans[st.top().ind] = i - st.top().ind;
                st.pop();
            }
            st.push(Pair(temp[i], i));
        }

        return ans;
    }
};

File: 739. Daily Temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] ans = new int[temperatures.length];
        Stack<int[]> st = new Stack<>();

        for(int i = 0; i < temperatures.length; i++){
            int currTemp = temperatures[i];
            while(!st.isEmpty() && st.peek()[0] < currTemp){
                int index = st.pop()[1];
                ans[index] = i - index;
            }
            st.push(new int[]{currTemp, i});
        }

        return ans;
    }
}